### PR TITLE
Support for automatic updates

### PR DESCRIPTION
**Pull request by ge0rdi**
I've returned back `Update` component.

It now fetches info about latest release on Github, informs user when an update is available (changelog is displayed) and allows to download latest installer by single click.

It is not quite easy to test.
One would need to compile `update.exe` with lower version than on Github (can be done by modifying default version in `version.props` and recompiling).

Also the change will become _visible_ only once users have version that does support updating and we will release new version (so basically +2 versions from now). So hopefully everything will work well at that point.

Here is how update dialog looks like now:
image

I'd also like to add support for _nightly_ updates (separate setting).

### CODE REVIEW

**Comment by XenHat** [comment]
I will test this later today, I currently am fixing partition hell on linux. :)

**Comment by XenHat** [comment]
I've been very busy with IRL. Let's just ship it and tackle errors as they arrive.

File: Src/Lib/DownloadHelper.cpp [hunk 1]
```diff
@@ -848,6 +745,85 @@ void VersionData::Swap( VersionData &data )
 	std::swap(languages,data.languages);
 }
 
+// very simple json "parser"
+// it can parse simple json values in form:
```
Comment (XenHat) [inline Src/Lib/DownloadHelper.cpp:749]: Impressive, but I would rather use a library in the future for added robustness. There are many minimal, extremely fast libraries available.

**Review by XenHat — APPROVED** [review]
Looks good.
Future roadmap: use a json library.

**Comment by ge0rdi** [comment]
> Future roadmap: use a json library.

I'm working also on nightly updates support.
And to parse data from AppVeyor I need a bit more sophisticated parser.
So I'm considering to use https://github.com/nlohmann/json.

**Comment by ge0rdi** [comment]
I have added support for nightly builds as well.

Since we don't have changelogs for nightly builds, I'm using commit messages to construct changelog.
It looks like this:
image

**Review by XenHat — APPROVED** [review]
That's great!

**Comment by NebelNidas** [comment]
Really glad this is finally working, but I have a suggestion: When an update is found, wouldn't it be better if  the `OK` button became the `Download` button? Because in the current implementation, the `OK` and `Cancel` button are kind of redundant, they close the window without having any affect. In my opinion, it should look like this:
Updater suggestion

**Comment by XenHat** [comment]
Being forced to update immediately goes against many people's grain.
a better suggestion would be :
[Update now] [Update later] [Ignore]

**Comment by ge0rdi** [comment]
@NebelNidas 
The dialog is exactly the same as in `Classic-Shell`. I've changed only updating mechanism (not how it looks like).
I guess we will improve it over time (as I have to admit it is really not up-to-date).
So thanks for your suggestions.

> Because in the current implementation, the OK and Cancel button are kind of redundant, they both close the window without having any affect.

The dialog currently combines two things (and I guess this is the cause of confusion):

- update settings
- update check and download

`OK` button actually stores settings (whether you want to check for updates automatically, whether you want nightlies).
`Cancel` just dismisses the dialog without storing anything.

> Being forced to update immediately goes against many people's grain.

I guess nobody is talking about _forcing_ anyone.

I think we should split the dialog into two, in first place.
One dialog for settings (where Ok/Cancel) buttons make sense.
And one for actual update check, where we can consider 3 buttons as you proposed Update now/Later/Ignore.

Or eventually 2 buttons `Update`/`Not now` plus checkbox for ignoring (because that is something that probably not many users will choose).

Also the other thing to consider is whether to offer actual upgrade, or just download (as it is now).
Upgrade is more convenient (and these days kind of standard).
But I'm not quite sure about executing something downloaded from the Internet without any kind of verification (since we don't sign installers atm, it is not easy to verify what was downloaded actually).

**Comment by XenHat** [comment]
> But I'm not quite sure about executing something downloaded from the Internet without any kind of verification (since we don't sign installers atm, it is not easy to verify what was downloaded actually).

very good point.

**Comment by Gaurav-Original-ClassicShellTester** [comment]
Thank you @ge0rdi. I finally forced myself to devote more time into checking out your great fixes to Open Shell. Still catching up.

**Pull request merged**